### PR TITLE
fix(task): unplan task

### PR DESCRIPTION
**Pull request by Rom1-B**
To remove the planning of a task, it was necessary to remove the whole task and recreate it without planning.

image

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes/no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !25166

### CODE REVIEW

File: front/commonitiltask.form.php [hunk 1]
```diff
@@ -93,6 +93,20 @@
...
+        "tracking",
+        //TRANS: %s is the user login
+        sprintf(__('%s updates a task'), $_SESSION["glpiname"])
+    );
+    $redirect = $itemtype::getFormURLWithID($task->getField($fk));
+    $handled = true;
+} else if (isset($_POST["unplan"])) {
+    $task->check($_POST["id"], UPDATE);
+    $task->unplan($_POST);
+
```
Comment (trasher) [inline front/commonitiltask.form.php:109]: Following log message should say "unplan" and not "updates", no? Looks good to me appart of that.

File: templates/components/itilobject/timeline/form_task.html.twig [hunk 2]
```diff
@@ -365,10 +365,16 @@
...
                      </script>
                      <div class="col-12">
-                        <button id="plan{{ rand }}" class="btn btn-outline-secondary d-block mb-2 text-truncate" onclick="showPlanUpdate{{ rand }}()" type="button">
+                        <button id="plan{{ rand }}" class="btn btn-outline-secondary mb-2 text-truncate" onclick="showPlanUpdate{{ rand }}()" type="button">
                            <i class="fas fa-calendar"></i>
                            <span>{{ __('Plan this task') }}</span>
                         </button>
+                        {% if subitem.can(id, constant('UPDATE')) %}
+                           <button class="btn btn-outline-danger me-2" type="submit" name="unplan"
+                                 onclick="return confirm('{{ __('Confirm the deletion of planning?') }}');">
```
Comment (trasher) [inline templates/components/itilobject/timeline/form_task.html.twig:374]: Is'nt there a better way than using `@onclick`?

Comment (orthagh) [reply templates/components/itilobject/timeline/form_task.html.twig:374]: for a confirm, i think it's the best way

File: front/commonitiltask.form.php [hunk 3]
```diff
@@ -113,7 +113,7 @@
         4,
         "tracking",
         //TRANS: %s is the user login
-        sprintf(__('%s updates a task'), $_SESSION["glpiname"])
+        sprintf(__('%s unplan a task'), $_SESSION["glpiname"])
```
Comment (trasher) [inline front/commonitiltask.form.php:116]:
```suggestion
        sprintf(__('%s unplans a task'), $_SESSION["glpiname"])
```
 I'm not entirely sure :)

File: templates/components/itilobject/timeline/form_task.html.twig (hunk 2 above)
Comment (trasher) [reply templates/components/itilobject/timeline/form_task.html.twig:374]: :+1:  then

**Review by trasher — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

File: src/CommonITILTask.php [hunk 4]
```diff
@@ -1831,4 +1831,20 @@ public function getInputFromVCalendar(VCalendar $vcalendar)
 
         return $input;
     }
+
+    public function unplan()
+    {
+        global $DB;
+        $DB->update(
```
Comment (cedric-anne) [inline src/CommonITILTask.php:1838]: You should use `$this->update()` method.

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by trasher — APPROVED** [review]

File: templates/components/itilobject/timeline/form_task.html.twig [hunk 5]
```diff
@@ -365,10 +365,17 @@
                         }
                      </script>
                      <div class="col-12">
-                        <button id="plan{{ rand }}" class="btn btn-outline-secondary d-block mb-2 text-truncate" onclick="showPlanUpdate{{ rand }}()" type="button">
+                        <button id="plan{{ rand }}" class="btn btn-outline-secondary text-truncate" onclick="showPlanUpdate{{ rand }}()" type="button">
                            <i class="fas fa-calendar"></i>
                            <span>{{ __('Plan this task') }}</span>
                         </button>
+                        {% if subitem.can(id, constant('UPDATE')) %}
```
Comment (cedric-anne) [inline templates/components/itilobject/timeline/form_task.html.twig:372]: 1. Button appears even if task is not planned. It should be visible only on already planned tasks.
2. When clicking on "Plan this task" button, the "unplan" button is still shown, maybe it should be hide.

**Review by cedric-anne — CHANGES_REQUESTED** [review]
Following warning is trigerred when unplanning a task:
```
glpiphplog.NOTICE:   *** PHP Notice (8): Undefined index: tickets_id in /var/www/glpi/src/CommonITILTask.php at line 405
  Backtrace :
  src/CommonDBTM.php:1719                            CommonITILTask->post_updateItem()
  src/CommonITILTask.php:1841                        CommonDBTM->update()
  front/commonitiltask.form.php:108                  CommonITILTask->unplan()
  front/tickettask.form.php:39                       include()
```

**Comment by cedric-anne** [comment]
> Following warning is trigerred when unplanning a task:
> 
> ```
> glpiphplog.NOTICE:   *** PHP Notice (8): Undefined index: tickets_id in /var/www/glpi/src/CommonITILTask.php at line 405
>   Backtrace :
>   src/CommonDBTM.php:1719                            CommonITILTask->post_updateItem()
>   src/CommonITILTask.php:1841                        CommonDBTM->update()
>   front/commonitiltask.form.php:108                  CommonITILTask->unplan()
>   front/tickettask.form.php:39                       include()
> ```

I added a commit to fix this. This was directly not related to your changes. `post_updateItem()` should always use `$this->fields`.

**Review by cedric-anne — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Pull request merged**